controller/controller_server/addons_server.py

```diff
@@ -0,0 +1,160 @@
+                mediator.delete_replication(replication.name)
+            else:
+                logger.info("idempotent case. replication is already deleted with system {}".format(other_system_id))
+
+        return pb2.DisableVolumeReplicationResponse()
+
+    @staticmethod
+    def _ensure_volume_role_for_replication(mediator, replication, is_to_promote):
+        if is_to_promote:
+            if replication.is_primary:
+                logger.info("idempotent case. volume is already primary")
+            else:
+                logger.info("promoting volume for replication {}".format(replication.name))
+                mediator.promote_replication_volume(replication.name)
+        else:
+            if replication.is_primary:
+                logger.info("demoting volume for replication {}".format(replication.name))
+                mediator.demote_replication_volume(replication.name)
+            else:
+                logger.info("idempotent case. volume is already secondary")
+
+    def _ensure_volume_role(self, request, context, is_to_promote, response_type):
+        set_current_thread_name(request.volume_id)
+        method_name = "PromoteVolume" if is_to_promote else "DemoteVolume"
+        logger.info(method_name)
+        utils.validate_addons_request(request)
+
+        volume_id_info = utils.get_volume_id_info(request.volume_id)
+        volume_internal_id = volume_id_info.internal_id
+
+        other_volume_id_info = utils.get_volume_id_info(request.target_volume_id)
+        other_volume_internal_id = other_volume_id_info.internal_id
+
+        other_system_id = request.parameters.get(config.PARAMETERS_SYSTEM_ID)
+
+        connection_info = utils.get_array_connection_info_from_secrets(request.secrets)
+        with get_agent(connection_info, volume_id_info.array_type).get_mediator() as mediator:
+            replication = mediator.get_replication(volume_internal_id, other_volume_internal_id, other_system_id)
+            if not replication:
+                message = "could not find replication for volume internal id: {} " \
+                          "with volume internal id: {} of system: {}".format(volume_internal_id,
+                                                                             other_volume_internal_id,
+                                                                             other_system_id)
+                return build_error_response(message, context, grpc.StatusCode.FAILED_PRECONDITION, response_type)
+            logger.info("found replication {} on system {}".format(replication.name, mediator.identifier))
+
+            self._ensure_volume_role_for_replication(mediator, replication, is_to_promote)
+
+        return response_type()
+
+    @handle_common_exceptions(pb2.PromoteVolumeResponse)
+    def PromoteVolume(self, request, context):
+        return self._ensure_volume_role(request, context, is_to_promote=True, response_type=pb2.PromoteVolumeResponse)
```

Comment:
`ensure` sound like validation, it's not really a good name for a method that promotes or demotes..

Reply:
`ensure` is not really validation - it means "make sure", i.e.:

1. be idempotent
2. fail if the state can't be reached